cpp/include/raft/selection/col_wise_sort.cuh

```diff
@@ -0,0 +1,346 @@
+ * @param out: output value(index) matrix
+ * @param n_rows: number rows of input matrix
+ * @param n_cols: number columns of input matrix
+ * @param bAllocWorkspace: check returned value, if true allocate workspace passed in workspaceSize
+ * @param workspacePtr: pointer to workspace memory
+ * @param workspaceSize: Size of workspace to be allocated
+ * @param stream: cuda stream to execute prim on
+ * @param sortedKeys: Optional, output matrix for sorted keys (input)
+ */
+        template <typename InType, typename OutType>
+        void sortColumnsPerRow(const InType* in,
+                               OutType* out,
+                               int n_rows,
+                               int n_columns,
+                               bool& bAllocWorkspace,
+                               void* workspacePtr,
+                               size_t& workspaceSize,
+                               cudaStream_t stream,
+                               InType* sortedKeys = nullptr)
+        {
+            // assume non-square row-major matrices
+            // current use-case: KNN, trustworthiness scores
+            // output : either sorted indices or sorted indices and input values
+            // future : this prim can be modified to be more generic and serve as a way to sort column entries
+            // per row
+            //          i.e. another output format: sorted values only
+
+            int totalElements          = n_rows * n_columns;
+            size_t perElementSmemUsage = sizeof(InType) + sizeof(OutType);
+            size_t memAlignWidth       = 256;
+
+            // @ToDo: Figure out dynamic shared memory for block sort kernel - better for volta and beyond
+            // int currDevice = 0, smemLimit = 0;
+            // CUDA_CHECK(cudaGetDevice(&currDevice));
+            // CUDA_CHECK(cudaDeviceGetAttribute(&smemLimit, cudaDevAttrMaxSharedMemoryPerBlock, currDevice));
+            // size_t maxElementsForBlockSort = smemLimit / perElementSmemUsage;
+
+            // for 48KB smem/block, can fit in 6144 4byte key-value pair
+            // assuming key-value sort for now - smem computation will change for value only sort
+            // dtype being size of key-value pair
+            std::map<size_t, int> dtypeToColumnMap = {{4, 12288},   // short + short
+                                                      {8, 12288},   // float/int + int/float
+                                                      {12, 6144},   // double + int/float
+                                                      {16, 6144}};  // double + double
+
+            if (dtypeToColumnMap.count(perElementSmemUsage) != 0 &&
+                n_columns <= dtypeToColumnMap[perElementSmemUsage]) {
+                // more elements per thread --> more register pressure
+                // 512(blockSize) * 8 elements per thread = 71 register / thread
+
+                // instantiate some kernel combinations
+                if (n_columns <= 512)
```

Comment:
I always get a little anxious about omitting braces in single-line ifs like this. It's too easy for another developer to come along later and accidentally introduce a bug by forgetting that the clauses need to be single-line. Can we add braces to these?